src/components/ColorPicker.tsx

```diff
@@ -42,14 +48,104 @@ const ColorSwatch: FC<{
+
+      let matchColor, match
+      // Get the colors and background colors used in current thought's value
+      const fgColors: Set<string> = new Set()
+      if (colorMatches) {
+        // colorMatches will be like this : [color="#ee82ee", color="#ff823e"] and match.slice(7, -1) will be #ee82ee
+        // If the thought is colored with many colors, matchColor will be null and if the thought is colored with one color, matchColor will be that color
+        colorMatches.forEach(match => fgColors.add(match.slice(7, -1)))
+        matchColor = fgColors.size > 1 ? null : fgColors.values().next().value
+      }
+
+      const bgColors: Set<string> = new Set()
+      while ((match = bgColorRegex.exec(currentThoughtValue)) !== null) if (match[1]) bgColors.add(match[1])
+      const matchBgColor = bgColors.size > 1 ? null : bgColors.values().next().value
+
+      return !!(
+        (textHexColor && textHexColor === (matchColor && addAlphaToHex(rgbToHex(matchColor)))) ||
+        (backHexColor && backHexColor === (matchBgColor && addAlphaToHex(rgbToHex(matchBgColor))))
+      )
+    }
+    return !!(
+      (textHexColor && textHexColor === selectedTextHexColor) ||
+      (backHexColor && backHexColor === selectedBackHexColor)
+    )
+  })
+
   /** Toggles the text color to the clicked swatch. */
   const toggleTextColor = (e: React.MouseEvent | React.TouchEvent) => {
     // stop toolbar button dip
     e.stopPropagation()
+    e.preventDefault()
+    console.log(color, backgroundColor, colors.bg, colors.fg)
+    if (backgroundColor || color !== 'default') dispatch(formatSelection('foreColor', color || colors.bg))
+    else dispatch(formatSelection('foreColor', colors.fg))
+    // Apply background color to the selection
+    if (backgroundColor && backgroundColor !== colors.bg)
+      dispatch(formatSelection('backColor', backgroundColor === 'inverse' ? colors.fg : backgroundColor))
+    else {
+      dispatch(formatSelection('backColor', colors.bg))
+      const thoughtContentEditable = document.querySelector(`[aria-label="editable-${currentThoughtId}"]`)
+      if (!thoughtContentEditable) return
+      // Delete the default background and foreground color from the current thought's innerHTML
+      const styledElements = thoughtContentEditable.querySelectorAll('[style]')
+      styledElements?.forEach(el => {
+        const elementColor = el.getAttribute('color')
+        const elementStyle = el.getAttribute('style')
+        if (
+          (elementColor && elementColor !== rgbToHex(colors.bg)) ||
+          (label === 'default' && elementStyle?.includes('background-color'))
+        )
+          el.removeAttribute('style')
+      })
+      dispatch((dispatch, getState) => {
+        const state = getState()
+        if (!state.cursor) return
+        const thought = getThoughtById(state, head(state.cursor))
+        const simplePath = simplifyPath(state, state.cursor)
+        // Edit the current thought with the changed innerHTML
```

Comment:
This says *what* the code does, but not *why*. Can you explain? (What problem does it solve? What would occur if this wasn't done?)